src/routes/Batch.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useEffect, useState, FormEvent } from "react";
+import { useSelector } from "react-redux";
+import { getBatchById } from "../store/batches/batches.selectors";
+import { useParams } from "react-router-dom";
+import { getProductById } from "../store/products/products.selectors";
+import {
+  useFetchBatchIngredients,
+  useAddBatchIngredient
+} from "../store/batch-ingredients/batch-ingredients.hooks";
+import { useFetchBatches } from "../store/batches/batches.hooks";
+import { useFetchProducts } from "../store/products/products.hooks";
+import { BatchIngredient } from "../store/batch-ingredients/batch-ingredients.types";
+import { getIngredientById } from "../store/ingredients/ingredients.selectors";
+import format from "date-fns/format";
+import { useFetchIngredients } from "../store/ingredients/ingredients.hooks";
+import IngredientSelect from "../components/IngredientSelect";
+import { IngredientId } from "../store/ingredients/ingredients.types";
+import { BatchId } from "../store/batches/batches.types";
+import ingredients from "../store/ingredients/ingredients.reducer";
+
+interface BatchIngredientListItemProps {
+  batchIngredient: BatchIngredient;
+}
+const BatchIngredientListItem: React.FC<BatchIngredientListItemProps> = props => {
+  const { batchIngredient } = props;
+  const ingredient = useSelector(
+    getIngredientById(batchIngredient.ingredient_id)
+  );
+  if (!ingredient) {
+    return null;
+  }
+
+  return (
+    <li>
+      {ingredient.name} ({ingredient.type}), {batchIngredient.quantity_kg}
+      {ingredient.unit}
+    </li>
+  );
+};
+
+interface AddBatchIngredientProps {
+  batch_id: BatchId;
+}
+const AddBatchIngredient: React.FC<AddBatchIngredientProps> = props => {
+  const { batch_id } = props;
+  const [ingredientId, setIngredientId] = useState<IngredientId | undefined>(undefined);
+  const [quantity, setQuantity] = useState<number | undefined>(undefined);
+
+  const { submit: fetchBatchIngredients } = useFetchBatchIngredients(batch_id);
+  const { submit: addBatchIngredient, status } = useAddBatchIngredient();
+
+  const ingredient = useSelector(getIngredientById(ingredientId || -1));
+
+  const onSubmit = (e: FormEvent) => {
+    e.preventDefault();
+
+    if (ingredientId && quantity) {
+      addBatchIngredient({
+        batch_id,
+        ingredient_id: ingredientId,
+        quantity_kg: quantity
+      });
+    }
+  };
+
+  useEffect(() => {
+    if (status === "SUCCESS") {
+      setIngredientId(undefined);
+      setQuantity(undefined);
```

Comment:
what is the meaning of 'undefined here'? It feels counter-intuitive - if the status is SUCCESS I'd naively expect these to be defined.

Reply:
So this is in the `<AddBatchIngredient />` component.
There are two bits of local state `ingredientId` and `quanitity`.
These bits of state control the value of the inputs for the `ingredient` and `quantity` of the new batch ingredient respectively.

After the ingredient/quantity tuple has been added to the batch, so the `status` (of the request) is "SUCCESS", we want to clear the input values, so they are ready to accept new values foe the next ingredient to be added

Reply:
And by setting their value to `undefined` we are clearing the input.
We could just as well set their value to be `""`, but i tend to be quite a fan of `undefined`s where possible

Reply:
makes total sense now cheers.